libc/bsearch.c

```diff
@@ -0,0 +1,213 @@
+
+
+/* compare function used by bsearch */
+static int node_cmp(const void *node1, const void *node2)
+{
+	return strcmp(((struct node *)node1)->string, ((struct node *)node2)->string);
+}
+
+
+TEST_GROUP(bsearch);
+
+
+TEST_SETUP(bsearch)
+{
+}
+
+TEST_TEAR_DOWN(bsearch)
+{
+}
+
+
+TEST(bsearch, zero_el)
+{
+	struct node *rval, node;
+	node.string = "adam";
+	node.length = strlen(node.string);
+
+	rval = (struct node *)bsearch((void *)&node, (void *)table,
+		0, sizeof(struct node), node_cmp);
+	TEST_ASSERT_NULL(rval);
+}
+
+
+TEST(bsearch, find_node_in)
+{
+	int i;
+	struct node *rval, node;
+
+	/* check for even number of elements */
+	for (i = 0; i < TABSIZE; i++) {
+		node.string = names[i];
+		node.length = strlen(names[i]);
+		rval = (struct node *)bsearch((void *)&node, (void *)table,
+			(size_t)TABSIZE, sizeof(struct node), node_cmp);
+		TEST_ASSERT_NOT_NULL(rval);
+		TEST_ASSERT_EQUAL_MEMORY(&table[i], rval, sizeof(*rval));
+	}
+
+	TEST_ASSERT_EQUAL_INT(1, check_tab(names));
+
+	/* check for odd number of elements */
+	for (i = 0; i < TABSIZE - 1; i++) {
+		node.string = names[i];
+		node.length = strlen(names[i]);
+		rval = (struct node *)bsearch((void *)&node, (void *)table,
+			(size_t)TABSIZE - 1, sizeof(struct node), node_cmp);
+		TEST_ASSERT_NOT_NULL(rval);
+		TEST_ASSERT_EQUAL_MEMORY(&table[i], rval, sizeof(*rval));
+	}
+
+	TEST_ASSERT_EQUAL_INT(1, check_tab(names));
+}
+
+
+TEST(bsearch, find_node_not_in)
+{
+	struct node *rval, node;
+	char name[8];
+
+	TEST_ASSERT_EQUAL_MEMORY(name, strcpy(name, "phoenix"), sizeof(name));
+	node.string = name;
+	node.length = strlen(name);
+	/* check for even number of elements */
+	rval = (struct node *)bsearch((void *)&node, (void *)table,
+		(size_t)TABSIZE, sizeof(struct node), node_cmp);
+	TEST_ASSERT_NULL(rval);
+
+	TEST_ASSERT_EQUAL_INT(1, check_tab(names));
+
+	/* check for odd number of elements */
+	rval = (struct node *)bsearch((void *)&node, (void *)table,
+		(size_t)TABSIZE - 1, sizeof(struct node), node_cmp);
+	TEST_ASSERT_NULL(rval);
+
+	TEST_ASSERT_EQUAL_INT(1, check_tab(names));
+}
+
+
+TEST(bsearch, not_sorted)
+{
+	struct node *rval, node;
+	char name[5];
+
+	TEST_ASSERT_EQUAL_MEMORY(name, strcpy(name, "adam"), sizeof(name));
+	node.string = name;
+	node.length = strlen(name);
+	/* check for even number of elements */
+	rval = (struct node *)bsearch((void *)&node, (void *)table,
+		(size_t)TABSIZE, sizeof(struct node), node_cmp);
+	TEST_ASSERT_NULL(rval);
+
+	TEST_ASSERT_EQUAL_INT(1, check_tab(names_nsort));
+
+	/* check for odd number of elements */
+	rval = (struct node *)bsearch((void *)&node, (void *)table,
+		(size_t)TABSIZE, sizeof(struct node), node_cmp);
+	TEST_ASSERT_NULL(rval);
+
+	TEST_ASSERT_EQUAL_INT(1, check_tab(names_nsort));
+}
+
+
+TEST_GROUP_RUNNER(bsearch)
+{
+	RUN_TEST_CASE(bsearch, zero_el);
+	prepare(names);
+	RUN_TEST_CASE(bsearch, find_node_in);
+	RUN_TEST_CASE(bsearch, find_node_not_in);
+	prepare(names_nsort);
+	RUN_TEST_CASE(bsearch, not_sorted);
```

Comment:
It's not checked:
![Screenshot from 2023-06-07 18-29-45](https://github.com/phoenix-rtos/phoenix-rtos-tests/assets/77304431/0a8c4a1c-f40d-425c-94c8-8111c9699062)

Two or more members comapre equal - please add such case.